File: code/python/Stack/dynamic/Stack.py
from typing import Union, Any

from Node import Node
from StructBase import StructBase
from Stack.interface import StackInterface


class Stack(StructBase, StackInterface):
    def __init__(self) -> None:
        super(StructBase, self)
        self.size = 0
        self._head: Node = None

    def __del__(self) -> None:
        self._head = None

    def __len__(self) -> int:
        return self.size

    def __str__(self) -> str:
        str_stack = "\tHEAD -> "
        str_stack += " -> ".join(str(e) for e in self.__iter__())
        str_stack += " -> NULL\n"
        return str_stack

    def __iter__(self) -> Any:
        self._tmpNode = self._head
        return self

    def __next__(self) -> Any:
        if self._tmpNode is None:
            raise StopIteration
        data = self._tmpNode
        self._tmpNode = data.link
        return data

    def is_empty(self) -> bool:
        return self._head is None

    def is_full(self) -> bool:
        return False

    def push(self, value: Union[int, Node]) -> None:
        if type(value) is 'int':
            value = Node(value)
        value.link = self._head
        self._head = value
        self.size += 1

    def pop(self) -> Union[int, Node]:
        try:
            if self.is_empty():
                raise Exception("Stack is empty.")
            node = self._head
            self._head = node.link
            self.size -= 1
            return node
        except Exception as exc:
            print(exc)

    def peek(self) -> Union[int, Node]:
        try:
            if self.is_empty():
                raise Exception("Stack is empty.")
            return self._head
        except Exception as exc:
            print(exc)

    @staticmethod
    def options():
        return ["Push", "Pop", "Show top", "Print stack", "Exit"]

File: code/typescript/Stack/dynamic/Stack.ts
import { Node } from "../../Node/Node";
import StructBase from "../../StructBase/StructBase";
import StackInterface from "../interface/StackInterface";

export default class Stack extends StructBase implements StackInterface {
  protected head: Node | undefined;

  static options = ["Push", "Pop", "Show top", "Print stack", "Exit"];

  constructor() {
    super();
    this.head = undefined;
  }

  *[Symbol.iterator]() {
    let tmpNode = this.head;
    let auxNode = this.head;
    while (tmpNode !== undefined) {
      auxNode = tmpNode;
      tmpNode = tmpNode.link;
      yield auxNode;
    }
  }

  isEmpty() {
    return !this.head;
  }

  isFull(): boolean {
    return false;
  }

  push(value: number | Node): void {
    if (typeof value === "number") value = { data: value, link: undefined };
    value.link = this.head;
    this.head = value;
  }

  pop(): number | Node | undefined {
    try {
      if (this.isEmpty()) throw new Error("Stack is empty.");
      const node = this.head;
      if (!node) return undefined;
      this.head = node.link;
      return node;
    } catch (err: any) {
      console.error;
    }
  }

  peek(): number | Node | undefined {
    try {
      if (this.isEmpty()) throw "Stack is empty. Please, push some value";
      return this.head;
    } catch (err: any) {
      console.error(err);
    }
  }

  toString(): string {
    let str = "\tHEAD -> ";
    str += [...this].map((elem) => elem.data).join(" -> ");
    str += this.isEmpty() ? "NULL\n" : " -> NULL\n";
    return str;
  }
}
